File: content/step3/tree/tools/show_object.py
#!/usr/bin/env python3

import zlib
import hashlib
from sys import argv
from re import compile

from resolve_ref import resolve_ref


def sha1_to_path(sha1, directory):
    prefix = sha1[:2]
    suffix = sha1[2:]
    path = '/'.join([directory, 'objects', prefix, suffix])
    return path


def hash(content):
    return hashlib.sha1(content).hexdigest()


def decompress(path):
    with open(path, 'rb') as compressed:
        return zlib.decompress(compressed.read())


def sha1_to_content(sha1, directory='.git'):
    decompressed = decompress(sha1_to_path(sha1, directory))
    assert sha1 == hash(decompressed)
    header_end = decompressed.find(b'\x00')
    header = decompressed[:header_end].decode()
    object_type = header.split(' ')[0]
    content = decompressed[header_end+1:]
    if object_type == 'tree':
        tree_entry = compile(br'(\d+) (.*?)\x00([\x00-\xff]{20})')
        entries = [
            '\t'.join((mode.decode().zfill(6), sha1.hex(), name.decode()))
            for (mode, name, sha1) in
            tree_entry.findall(content)
        ]
        return [header, *entries]
    elif object_type in ('commit', 'blob'):
        return [header, content.decode()]

if __name__ == "__main__":
    ref = resolve_ref(argv[1], '.git')
    content = sha1_to_content(ref, '.git')[1:]
    [print(item) for item in content]

File: content/step3/tree/tools/ShowObject.hs
#!/usr/bin/env stack
-- stack --resolver lts-9.0 script

{-# LANGUAGE OverloadedStrings #-}

module Main where

import qualified Codec.Compression.Zlib           as Z (decompress)
import           Control.Monad
import           Data.Attoparsec.ByteString.Char8 as A
import           Data.ByteString                       (ByteString)
import qualified Data.ByteString                  as B
import           Data.ByteString.Base16                (encode)
import qualified Data.ByteString.Char8            as C
import           Data.ByteString.Lazy                  (fromStrict, toStrict)
import           Data.ByteString.UTF8                  (fromString, toString)
import           Data.Digest.Pure.SHA
import           Data.List
import           GHC.Word                              (Word8)
import           System.Environment
import           System.FilePath
import           System.Directory

sha1ToDirectory :: String -> FilePath -> FilePath
sha1ToDirectory (x:y:_) dir = dir </> "objects" </> [x,y]

sha1ToPath :: String -> FilePath -> FilePath
sha1ToPath (x:y:suffix) dir = dir </> "objects" </> [x,y] </> suffix

disambiguateSha1 :: String -> FilePath -> IO String
disambiguateSha1 sha1@(x:x':suffix) dir = do
    let prefix    =  [x,x']
    let searchDir =  sha1ToDirectory sha1 dir
    matchingFiles <- filter (isPrefixOf suffix) <$> listDirectory searchDir
    case length matchingFiles `compare` 1 of
        LT -> error "no object exists with that SHA1"
        GT -> error "ambiguous SHA1 provided"
        EQ -> return $ prefix ++ head matchingFiles

readRef :: FilePath -> IO String
readRef = fmap init . readFile

resolveRef :: String -> FilePath -> IO String
resolveRef ref dir = do
    let symRefPath      =  dir </> ref
    symRefPathExists    <- doesFileExist symRefPath
    let refsHeadsPath   =  dir </> "refs" </> "heads" </> ref
    refsHeadsPathExists <- doesFileExist refsHeadsPath
    let refsTagsPath    =  dir </> "refs" </> "tags"  </> ref
    refsTagsPathExists  <- doesFileExist refsTagsPath
    case () of
      _ | "ref" `isPrefixOf` ref -> readRef symRefPath
        | symRefPathExists       -> do
            content <- readRef symRefPath
            if "ref: " `isPrefixOf` content
                then resolveRef (drop 5 content) dir
                else return content
        | refsHeadsPathExists    -> readRef refsHeadsPath
        | refsTagsPathExists     -> readRef refsTagsPath
        | otherwise              -> disambiguateSha1 ref dir

hash :: ByteString -> ByteString
hash = fromString . showDigest . sha1 . fromStrict

decompress :: ByteString -> ByteString
decompress = toStrict . Z.decompress . fromStrict

parseTreeEntry :: Parser [ByteString]
parseTreeEntry = do
    perms <- fromString <$> A.many' A.digit
    A.space
    name  <- A.takeWhile (/= '\NUL')
    A.char '\NUL'
    ref   <- encode <$> A.take 20
    return [perms, name, ref]

parsedObject :: ByteString -> [ByteString]
parsedObject raw = let
    (headerLen, rest) = C.break (=='\NUL') raw
    content           = B.tail rest
    objectType        = fst $ C.break (==' ') headerLen
    in case objectType of
        "tree" -> let
                parser = A.many' parseTreeEntry
                parsed = either error id $ A.parseOnly parser content
            in map (B.intercalate "\t") parsed
        _ -> [content]

main :: IO ()
main = do
    (ref':_) <- getArgs
    ref      <- resolveRef ref' ".git"
    let path =  sha1ToPath ref  ".git"
    content  <- decompress <$> B.readFile path
    traverse C.putStrLn (parsedObject content)
    unless (fromString ref == hash content) $ error "refs don't match"
